docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,188 @@
+1. Translating Cluster API resource definitions into Azure infrastructure definitions
+2. Interfacing with the Azure platform to manage creating, updating, and deleting that infrastructure
+
+Currently, CAPZ handles `2.` with Azure API calls, which requires ensuring dependencies between Azure resources are met and maintaining state to asynchronously handle long-running operations. Because ASO's primary objective is to handle `2.`, CAPZ could then more declaratively specify all of the infrastructure needed for a given Cluster API resource and let ASO handle the finer details of ensuring that desired state is achieved in Azure. By then primarily focusing on `1.`, CAPZ reduces its overall code maintenance burden and affirms ASO as the canonical way to manage Azure infrastructure from Kubernetes.
+
+In addition, the [Azure Go SDK](https://github.com/Azure/azure-sdk-for-go) has recently deprecated its set of APIs that CAPZ currently uses in favor of a new set of backwards-incompatible APIs. Because the deprecated Azure Go SDK APIs are no longer receiving updates to adopt new Azure API versions, CAPZ must adopt the new APIs to be able to expose the newest features in Azure to its users. Because ASO already uses the new APIs, adopting ASO in CAPZ accomplishes the goal of migrating off of the deprecated APIs.
+
+### Goals
+
+- Logically separate transformation of Cluster API resources to infrastructure definitions from reconciliation of that infrastructure in Azure
+- Determine a method to migrate CAPZ's individual Azure service interfaces one-by-one to use ASO
+- Determine how ASO will be installed on Cluster API management clusters
+
+### Non-Goals/Future Work
+
+- Entirely eliminate direct usage of any Azure Go SDK
+- Reconcile long-running Azure operations asynchronously by waiting for
+  Kubernetes watch events from changes to the `status` of ASO resources instead
+  of periodically checking if the operation is complete
+- Reconcile different Azure resource types for the same workload cluster in parallel
+- Allow users to interact with ASO resources directly, out-of-band with CAPZ
+- Adopt Azure Workload Identity to authenticate with Azure APIs
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Users of new Azure features
+
+As a CAPZ user I want to take advantage of the latest and greatest features in Azure only available in the newest Azure API versions in my clusters managed by CAPZ. This would allow me to adopt CAPZ for a wider range of use cases and continue evolving my CAPZ clusters alongside the Azure API.
+
+#### Story 2 - Developer experience
+
+As a CAPZ developer I would like to only need to interact with the Kubernetes API server using Custom Resource Definitions representing Azure resources rather than having to deal with the impedance of interacting with the Azure APIs. This would enable me to focus more on the core value of CAPZ, transforming Cluster API resource definitions to Azure resource definitions.
```

Comment:
How about adding something additional WRT making it easier to develop CAPZ features.  e.g. "I would like to have to write minimal code to adopt new features available in the API."  Providing code sample comparisons would be a bonus -  something like provisioning a resource group in CAPZ with Go SDK versus ASO.

Reply:
I'm not actually all that confident this will make adding features any easier since ultimately CAPZ would still be constructing a resource to submit for creation/updating, and the SDK and ASO resources are structured almost identically.

Here's an example of the entire SDK-facing part of the changes required to add a new feature from #3046:
```diff
diff --git a/azure/services/managedclusters/spec.go b/azure/services/managedclusters/spec.go
index a88c55d5cb..9ebf6f0114 100644
--- a/azure/services/managedclusters/spec.go
+++ b/azure/services/managedclusters/spec.go
@@ -65,6 +65,9 @@ type ManagedClusterSpec struct {
 	// NetworkPolicy used for building Kubernetes network. Possible values include: 'calico', 'azure'.
 	NetworkPolicy string
 
+	// OutboundType used for building Kubernetes network. Possible values include: 'loadBalancer', 'managedNATGateway', 'userAssignedNATGateway', 'userDefinedRouting'.
+	OutboundType *infrav1.ManagedControlPlaneOutboundType
+
 	// SSHPublicKey is a string literal containing an ssh public key. Will autogenerate and discard if not provided.
 	SSHPublicKey string
 
@@ -377,6 +380,10 @@ func (s *ManagedClusterSpec) Parameters(ctx context.Context, existing interface{
 		}
 	}
 
+	if s.OutboundType != nil {
+		managedCluster.NetworkProfile.OutboundType = containerservice.OutboundType(*s.OutboundType)
+	}
+
 	managedCluster.AutoScalerProfile = buildAutoScalerProfile(s.AutoScalerProfile)
 
 	if existing != nil {
```

I think the changes with ASO in the mix would look similar overall.